### PR TITLE
Implement Galileo CBOC

### PR DESCRIPTION
**Pull request by zsoerenm**
This uses a different approach to implement the modulation. It uses Holy-Traits to implement the modulation.
Watch out for `get_modulation` e.g.:
```julia
get_modulation(::Type{<:GalileoE1B}) = CBOC(BOCsin(1, 1), BOCsin(6, 1), 10 / 11)
```
This fixes https://github.com/JuliaGNSS/GNSSSignals.jl/issues/25

### CODE REVIEW

**Comment by zsoerenm** [comment]
@dominformant I think this ready for review.

**Comment by dominformant** [comment]
> @dominformant I think this ready for review.

This looks quite nice, however I'm quite busy at the moment, but will look deeper into it as soon as possible.

**Comment by zsoerenm** [comment]
@dominformant Did you have a change to look at it? ;)

File: src/modulation.jl [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    CBOC(boc1, boc2, boc1_power) =
+        0 < boc1_power < 1 && boc1.n == boc2.n ?
+        new{typeof(boc1), typeof(boc2)}(boc1, boc2, boc1_power) :
+        error("Power of BOC1 must be between 0 and 1 and n of both BOCs must match")
+end
+
+get_code_type(system::T) where T <: AbstractGNSS = get_code_type(system, get_modulation(T))
+
+get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation) where T = T
+get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation::CBOC) where T = typeof(modulation.boc1_power)
```
Comment (dominformant) [inline src/modulation.jl:37]: I think here we should use type promotion to be save, even if the type of the PRN matrix is more general than the type of the power ratio.
```suggestion
get_code_type(system::AbstractGNSS{<:AbstractMatrix{T}}, modulation::CBOC) where T = promote_type(T, typeof(modulation.boc1_power))
```

File: src/modulation.jl [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+$(SIGNATURES)
+Get the spectral power
+"""
+get_code_spectrum(system::T, f) where T <: AbstractGNSS = get_code_spectrum(get_modulation(T), system, f)
+get_code_spectrum(modulation::LOC, system, f) = get_code_spectrum_BPSK(get_code_frequency(system), f)
+get_code_spectrum(modulation::BOCsin, system, f) =
+    get_code_spectrum_BOCsin(modulation.n * get_code_frequency(system), modulation.m * get_code_frequency(system), f)
+get_code_spectrum(modulation::BOCcos, system, f) =
+    get_code_spectrum_BOCcos(modulation.n * get_code_frequency(system), modulation.m * get_code_frequency(system), f)
+get_code_spectrum(modulation::CBOC, system, f) = error("To be implemented")
```
Comment (dominformant) [inline src/modulation.jl:54]: I think this should be
```suggestion
get_code_spectrum(modulation::CBOC, system, f) = get_code_spectrum(m.boc1, system, f) * m.boc1_power + get_code_spectrum(m.boc2, system, f) * (1-m.boc1_power)
```

**Review by dominformant — COMMENTED** [review]
Just some small ideas, the rest looks good

**Pull request merged**